src/auth/baseexternalclient.ts

```diff
@@ -0,0 +1,487 @@
+   * { Authorization: 'Bearer <access_token_value>' }
+   */
+  async getRequestHeaders(): Promise<Headers> {
+    const accessTokenResponse = await this.getAccessToken();
+    const headers: Headers = {
+      Authorization: `Bearer ${accessTokenResponse.token}`,
+    };
+    return this.addSharedMetadataHeaders(headers);
+  }
+
+  /**
+   * Provides a request implementation with OAuth 2.0 flow. In cases of
+   * HTTP 401 and 403 responses, it automatically asks for a new access token
+   * and replays the unsuccessful request.
+   * @param opts Request options.
+   * @param callback callback.
+   * @return A promise that resolves with the HTTP response when no callback is
+   *   provided.
+   */
+  request<T>(opts: GaxiosOptions): GaxiosPromise<T>;
+  request<T>(opts: GaxiosOptions, callback: BodyResponseCallback<T>): void;
+  request<T>(
+    opts: GaxiosOptions,
+    callback?: BodyResponseCallback<T>
+  ): GaxiosPromise<T> | void {
+    if (callback) {
+      this.requestAsync<T>(opts).then(
+        r => callback(null, r),
+        e => {
+          return callback(e, e.response);
+        }
+      );
+    } else {
+      return this.requestAsync<T>(opts);
+    }
+  }
+
+  /**
+   * @return A promise that resolves with the project ID corresponding to the
+   *   current workload identity pool. When not determinable, this resolves with
+   *   null.
+   *   This is introduced to match the current pattern of using the Auth
+   *   library:
+   *   const projectId = await auth.getProjectId();
+   *   const url = `https://dns.googleapis.com/dns/v1/projects/${projectId}`;
+   *   const res = await client.request({ url });
+   *   The resource may not have permission
+   *   (resourcemanager.projects.get) to call this API or the required
+   *   scopes may not be selected:
+   *   https://cloud.google.com/resource-manager/reference/rest/v1/projects/get#authorization-scopes
+   */
+  async getProjectId(): Promise<string | null> {
+    if (this.projectId) {
+      // Return previously determined project ID.
+      return this.projectId;
+    } else if (this.projectNumber) {
+      // Preferable not to use request() to avoid retrial policies.
+      const headers = await this.getRequestHeaders();
+      const response = await this.transporter.request<ProjectInfo>({
+        headers,
+        url: `${CLOUD_RESOURCE_MANAGER}${this.projectNumber}`,
+        responseType: 'json',
+      });
+      this.projectId = response.data.projectId;
+      return this.projectId;
+    }
+    return null;
+  }
+
+  /**
+   * Authenticates the provided HTTP request, processes it and resolves with the
+   * returned response.
+   * @param opts The HTTP request options.
+   * @param retry Whether the current attempt is a retry after a failed attempt.
+   * @return A promise that resolves with the successful response.
+   */
+  protected async requestAsync<T>(
+    opts: GaxiosOptions,
+    retry = false
+  ): Promise<GaxiosResponse<T>> {
+    let r2: GaxiosResponse;
```

Comment:
[readability] Can we fix the naming? `r` and `r2` don't read good enough.

Reply:
Done